tutorial/chapter1.md

```diff
@@ -1,41 +1,210 @@
+```cpp
+    serviceRunnerThreads[i].join();
+    readFromDiskServices[i]->shutdown();
+    subscribers[i]->removeListener(prefixChannel + std::to_string(i));
+eventLoopProvider.stop();
+```
+
+Typical output looks like:
+
+```text
+[2023-09-04 13:59:47.280] [info] Read a list of 5 images in folder: tests/data
+[2023-09-04 13:59:47.280] [info] Read a list of 5 images in folder: tests/data
+[2023-09-04 13:59:47.280] [info] Read a list of 5 images in folder: tests/data
+[2023-09-04 13:59:47.280] [info] Service Runner 0. Iteration 0. Thread-id: 139991226636032
+[2023-09-04 13:59:47.280] [info] Service Runner 1. Iteration 0. Thread-id: 139991218243328
+[2023-09-04 13:59:47.280] [info] Service Runner 2. Iteration 0. Thread-id: 139991209850624
+[2023-09-04 13:59:47.466] [info] Service Runner 1. Iteration 1. Thread-id: 139991218243328
+[2023-09-04 13:59:47.469] [info] Listener 1. Thread-id: 139991310563072
+[2023-09-04 13:59:47.514] [info] Service Runner 2. Iteration 1. Thread-id: 139991209850624
+[2023-09-04 13:59:47.517] [info] Listener 2. Thread-id: 139991310563072
+[2023-09-04 13:59:47.528] [info] Listener 0. Thread-id: 139991310563072
+[...]
+[2023-09-04 13:59:48.406] [info] Service Runner 1. Iteration 5. Thread-id: 139991218243328
+[2023-09-04 13:59:48.406] [info] Read a list of 5 images in folder: tests/data
+[2023-09-04 13:59:48.422] [info] Listener 2. Thread-id: 139991310563072
+[2023-09-04 13:59:48.496] [info] Listener 0. Thread-id: 139991310563072
+[2023-09-04 13:59:48.600] [info] Listener 1. Thread-id: 139991310563072
+[2023-09-04 13:59:48.622] [info] Total frames listened: 18.
+```
 
-## Core API example <a name="core-api-example"></a>
+## Event loop example 2
 
-The [mem example](../examples/mem_example/src/main.cpp) shows the usage of the [`DataMultiplexerMiddlewareProvider`](https://github.com/klepsydra-technologies/kpsr-core/blob/main/high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h) high performance API (from kpsr-core) when processing opencv `Mat` images using `ImageDataFactory` and `ImageMultiplexerVisionMiddlewareProvider` classes. In this sample example, we process images by resizing them to desired size and publish them. After that, the mean of these resized images is calculated and published too.
+This example is similar to example 1 with some differences:
 
-Execute this example using the command `bin/kpsr_image_processing_mem_example` to process the images saved in `../tests/data` directoy, but if you prefer to process other images execute it providing any images path.
+- There are 2 cameras: each camera has its own even loop and publisher.
+- One additional publisher for the image transformation.
+- Each camera has two subscribers: one for the images reading and one for transformation, with the help of `kpsr::EventTransformForwarder`:
```

Comment:
Advantages of `kpsr::EventTransformForwarder`?

Reply:
With `EventTransformForwarder` the user does not need to manage the creation of the event that would be forwarded to the next step.

The user also does not need to manually publish the event to the next layer.

The code is cleaner and more akin to the [map pattern](https://en.wikipedia.org/wiki/Map_(parallel_pattern)).

Reply:
Explanation added.